test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/net/URLEncoderTest.scala

```diff
@@ -0,0 +1,106 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package org.scalajs.testsuite.javalib.net
+
+import org.junit.Test
+import org.junit.Assert._
+
+import org.scalajs.testsuite.utils.AssertThrows.assertThrows
+import org.scalajs.testsuite.utils.Platform._
+
+import java.net.URLEncoder
+import java.io.UnsupportedEncodingException
+
+class URLEncoderTest {
+  private final val utf8 = "utf-8"
+
+  @Test
+  def encodeNoCharset(): Unit = {
+    def test(unencoded: String, expected: String): Unit =
+      assertEquals(expected, URLEncoder.encode(unencoded))
+
+    // empty string
+    test("", "")
+
+    // . - * _ remain the same, as well as ASCII letters and digits
+    test("afz-AFZ-069-.*_", "afz-AFZ-069-.*_")
+
+    // ' ' -> '+'
+    test("a b c", "a+b+c")
+
+    // single byte codepoint
+    test("a@b#c", "a%40b%23c")
+
+    // multi byte codepoint, include surrogate pairs
+    test("aßcd\uD834\uDD1Eé", "a%C3%9Fcd%F0%9D%84%9E%C3%A9") // 𝄞 U+1D11E MUSICAL SYMBOL G CLEF
+
+    // consecutive characters
+    test("aß#c", "a%C3%9F%23c")
+  }
+
+  @Test
+  def encodeStringCharset(): Unit = {
+    def test(unencoded: String, expected: String, enc: String = utf8): Unit =
+      assertEquals(expected, URLEncoder.encode(unencoded, enc))
+
+    def unsupportedEncoding(unencoded: String, enc: String = utf8): Unit = {
+      val exception = classOf[UnsupportedEncodingException]
+      assertThrows(exception, URLEncoder.encode(unencoded, enc))
+    }
+
+    // empty string
+    test("", "")
+
+    // . - * _ remain the same, as well as ASCII letters and digits
+    test("afz-AFZ-069-.*_", "afz-AFZ-069-.*_")
+
+    // ' ' -> '+'
+    test("a b c", "a+b+c")
+
+    // single byte codepoint
+    test("a@b#c", "a%40b%23c")
+
+    // multi byte codepoint, include surrogate pairs
+    test("aßcd\uD834\uDD1Eé", "a%C3%9Fcd%F0%9D%84%9E%C3%A9") // 𝄞 U+1D11E MUSICAL SYMBOL G CLEF
+
+    // consecutive characters
+    test("aß#c", "a%C3%9F%23c")
+
+    // invalid encoding
+    unsupportedEncoding("2a%20b%20c", enc = "dummy")
+
+    /* Throw even if the charset is not needed.
+     * Despite what the documentation says, `decode` eagerly throws when the
+     * charset is not supported. This behavior started in JDK 10, when they
+     * added the overload of `decode` taking a `Charset`. With that addition,
```

Comment:
Encode instead of decode?

Reply:
Oh yes I forgot this. Actually it's more than that: for `URLEncoder` the eager throw behavior has always been specified, and even JDK 8 does it.